gateway/main.go

```diff
@@ -1,51 +1,185 @@
 package main
 
 import (
-	"log"
+	"context"
+	"time"
+
+	"github.com/shellhub-io/shellhub/pkg/envs"
+	"github.com/shellhub-io/shellhub/pkg/loglevel"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// defaultNginxRootDir is the default base directory for Nginx configuration files.
+	// This directory typically contains the main nginx.conf and site configuration files.
+	defaultNginxRootDir = "/etc/nginx"
+
+	// defaultNginxTemplateDir is the default directory where Nginx template files are stored.
+	// Template files are used to generate dynamic Nginx configurations based on runtime settings.
+	defaultNginxTemplateDir = "/templates"
+
+	// defaultCertBotRootDir is the default directory where Certbot keeps
+	// generated certificates, keys, and related assets.
+	// This follows the standard Let's Encrypt directory structure.
+	defaultCertBotRootDir = "/etc/letsencrypt"
+
+	// defaultTickerRenewCertificates defines the interval for automatic certificate renewal checks.
+	// Certificates are checked for renewal every 24 hours to ensure they remain valid.
+	defaultTickerRenewCertificates = 24 * time.Hour
 )
 
+const (
+	// NoFeature indicates no feature is enabled.
+	NoFeature = 0
+	// SSLFeature indicates that SSL's feature is eanbled
+	SSLFeature = 1 << iota
+	// TunnelsFeature indicates that Tunnels' feature is eanbled.
+	TunnelsFeature
+)
+
+// Gateway represents the main gateway service that orchestrates Nginx configuration
+// management and SSL certificate provisioning.
+type Gateway struct {
+	// Config holds the gateway's configuration settings including domain,
+	// environment, and SSL settings.
+	Config *GatewayConfig
+
+	// Controller manages Nginx configuration generation, template processing,
+	// and server lifecycle operations.
+	Controller *NginxController
+
+	// Certbot handles SSL certificate provisioning and renewal through Let's Encrypt.
+	// This field is nil when SSL is not enabled.
+	Certbot *CertBot
+
+	// Features contains feature flags to gateway.
+	Features int
+}
+
+// NewGateway creates a new Gateway instance with the provided configuration and controller.
+// The Certbot component is initially set to nil and will be initialized only when
+// SSL is explicitly enabled through EnableSSL().
+func NewGateway(config *GatewayConfig, controller *NginxController, features int) *Gateway {
+	g := &Gateway{
+		Config:     config,
+		Controller: controller,
+		Certbot:    nil,
+	}
+
+	// NOTE: [SSLFeature] indicates that SSL's feature is eanbled, configuring SSL certificate management for the
+	// gateway. It sets up Certbot with the gateway's domain configuration and establishes automatic certificate
+	// provisioning and renewal. The renewal callback is configured to reload Nginx when certificates are
+	// renewed, ensuring the server uses the latest certificates without manual intervention.
+	if features&SSLFeature != 0 {
```

Comment:
I understand that we are using a bit flag here, but I don't see too much value in save few bytes here, and instead I would propose that we use a feature list, where you can add one or more features, and if no feature is enabled, the list just needs to be empty, and that would also allow you to check if it contains a feature inside of the list, and you could rely on constants for the features to be added on the list for later comparison. I think that it makes the code clear and avoids some potential mistakes.